authoring-recipes/multiple-visitors.md

```diff
@@ -0,0 +1,814 @@
+
+{% hint style="note" %}
+The `VariableDeclarations` LST contains all the variables defined on one line of code. In most cases, this will just be a single variable, but it is possible for many to be defined. Our visitor will need to handle both cases.
+{% endhint %}
+
+```java
+// ...
+public class FinalizeLocalVariables extends Recipe {
+    // ...
+
+    @Override
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations varDec, ExecutionContext executionContext) {
+                J.VariableDeclarations variableDeclarations = super.visitVariableDeclarations(varDec, executionContext);
+
+                // TODO: Return if the varDec already has the final modifier
+
+                // TODO: Return if any variables in the varDec are uninitialized
+
+                // TODO: Return if the varDec is in a for loop control
+
+                // TODO: Return if the varDec is an instance or class variable
+
+                // TODO: If none of the above apply, find all use cases for the variables and determine whether or not they're reassigned. If none have been, add the final modifier. If one ore more has, do nothing.
+
+                return variableDeclarations;
+            }
+        };
+    }
+}
+```
+
+{% hint style="warning" %}
+In most cases, your recipe should call `super.visitSomeLST()` at the beginning of the overridden method. This allows OpenRewrite to traverse into sub-elements of the LST you're visiting. In the above recipe, if we didn't call super, the recipe would fail to add the `final` modifier to the `c` variable in [this code](https://gist.github.com/mike-solomon/cdb8cb02a64f016ef225bf03a4d25bd1) as it would never visit the sub-elements of `a`.
+{% endhint %}
+
+When implementing the body of the visitor, you may find that the LST you're working with has all of the relevant information contained in it. For instance, the first two tasks in the above visitor can be implemented without diving into other LSTs:
+
+```java
+@Override
+public JavaIsoVisitor<ExecutionContext> getVisitor() {
+    return new JavaIsoVisitor<ExecutionContext>() {
+
+        @Override
+        public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations varDec, ExecutionContext executionContext) {
+            J.VariableDeclarations variableDeclarations = super.visitVariableDeclarations(varDec, executionContext);
+            
+            // If the varDec already has a "final" modifier, we don't need to check anything else
+            if (variableDeclarations.hasModifier(J.Modifier.Type.Final)) {
+                return variableDeclarations;
+            }
+
+            // Uninitialized local variables should not be considered
+            if (variableDeclarations.getVariables().stream().anyMatch(namedVariable -> namedVariable.getInitializer() == null)) {
```

Comment:
This is not quite true and our recipe could be improved to finalize more local variables. If the compiler can determine that a final variable is assigned exactly once before it is first dereferenced, it doesn't have to be assigned using an initializer. So the following is also perfectly valid:
```java
final int x;
if (...) {
  x = 3;
} else {
  x = 4;
}
```
We could enhance our recipe to also add the `final` modifier in these cases. But I am wondering if it is worth the trouble.

I just wanted to point this out, in case we want to change the comment to point out that we intentionally ignore these more complex cases.

Reply:
Thanks for the feedback @knutwannheden . I know it's possible to expand on this recipe to handle more cases (and I think @kunli2 is working on that). However, I wanted to limit the scope of the recipe to mostly what exists in it right now. 

In the scope section at the beginning of the document, I call out that for the purposes of this guide, we won't consider uninitialized variables: https://github.com/openrewrite/rewrite-docs/blob/8fa687ce1024816ab34fe9bf69532a4274f729cd/authoring-recipes/multiple-visitors.md?plain=1#L42

Do you think it would help if I re-iterate that here? Like change the comment to be `// Uninitialized local variables should not be considered in this tutorial`?

Reply:
No, I missed that part in the scope. Thanks for clarifying!

Reply:
Just FYI, I am going to improve the recipe `FinalizeLocalVariables` to cover more case (include without initializer), but for the case of `if / else` as Knut exampled above.
```java
final int x;
if (...) {
  x = 3;
} else {
  x = 4;
}
```
I haven't got a solution for this yet, so we will leave it uncovered.